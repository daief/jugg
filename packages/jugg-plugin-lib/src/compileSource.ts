// import gulp from 'gulp';

// export default (api: Jugg) => {
//   gulp.
// }
