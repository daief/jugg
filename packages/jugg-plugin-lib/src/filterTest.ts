import through2 from 'through2';

export function filterTest() {
  return through2.obj((file, _, callback) => {
    const filename = file.path;
    if (/(\/__test__\/)|(\/test\/)|(\.test\.)|(\.spec\.)/gi.test(filename)) {
      return callback(null, null);
    }
    callback(null, file);
  });
}

export default filterTest;
