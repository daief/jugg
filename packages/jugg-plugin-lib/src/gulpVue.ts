/*
 * @Author: daief
 * @LastEditors: daief
 * @Date: 2019-08-15 23:53:23
 * @Description:
 */
import { PluginAPI } from '@axew/jugg/types/PluginAPI';
import {
  // assemble,
  AssembleOptions,
  // createDefaultCompiler,
  ScriptOptions,
  StyleOptions,
  TemplateOptions,
} from '@vue/component-compiler';
import path from 'path';
import through2 from 'through2';
import * as ts from 'typescript';

let isWarned = false;

export interface IOptions {
  compilerOptions?: {
    script?: ScriptOptions | undefined;
    style?: StyleOptions | undefined;
    template?: TemplateOptions | undefined;
  };
  assembleOptions?: AssembleOptions;
  // config file
  tsconfig?: string;
}

export function compileVueFile(
  content: string,
  filename: string,
  opts: IOptions = {},
) {
  const {
    compilerOptions,
    assembleOptions,
    // tsconfig = ''
  } = opts;
  const {
    assemble,
    createDefaultCompiler,
  } = require('@vue/component-compiler');

  // const { parse } = require('@vue/component-compiler-utils');
  // const templateCompiler = require('vue-template-compiler');
  // const _descriptor = parse({
  //   source: content,
  //   filename,
  //   needMap: true,
  //   compiler: templateCompiler,
  // });

  const compiler = createDefaultCompiler(compilerOptions);
  const descriptor = compiler.compileToDescriptor(filename, content);

  // TS to js
  descriptor.script.code = ts.transpile(
    descriptor.script.code,
    {
      // TODO 合并 ts 配置
      target: ts.ScriptTarget.ES2015, // stay es module
    },
    filename,
  );

  // 这里只能处理 js
  const result = assemble(compiler, filename, descriptor, {
    normalizer: '~' + 'vue-runtime-helpers/dist/normalize-component.js',
    styleInjector: '~' + 'vue-runtime-helpers/dist/inject-style/browser.js',
    styleInjectorSSR: '~' + 'vue-runtime-helpers/dist/inject-style/server.js',
    ...assembleOptions,
  });

  return result.code;
}

/**
 * convert signle `vue` to `ts`
 * @param opts
 * @param api
 */
export function gulpVue(api: PluginAPI, opts: IOptions = {}) {
  if (!api.jugg.Utils.resolveCwd.silent('vue-template-compiler')) {
    !isWarned &&
      api.jugg.Utils.logger.warn(
        '`.vue` files will not be resolved. Install `vue-template-compiler` to support.',
        '[jugg-plugin-lib]',
      );
    isWarned = true;
    return through2.obj((file, __, callback) => callback(file));
  }

  return through2.obj((file, __, callback) => {
    const filename = path.basename(file.path);

    const contentStr = file.contents.toString();

    file.path = file.path.replace(/\.vue$/, '.ts');

    file.contents = Buffer.from(compileVueFile(contentStr, filename, opts));

    callback(null, file);
  });
}

export default gulpVue;
