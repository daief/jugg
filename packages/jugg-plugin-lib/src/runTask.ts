import gulp from 'gulp';

// ref: https://github.com/ant-design/antd-tools/blob/b28fd3685406d06dabb8e0cc8ffe4603336aab69/lib/cli/run.js#L16
export function runTask(toRun: string) {
  const metadata: any = { task: toRun };
  // Gulp >= 4.0.0 (doesn't support events)
  const taskInstance = gulp.task(toRun);
  if (taskInstance === undefined) {
    gulp.emit('task_not_found', metadata);
    return;
  }
  const start = process.hrtime();
  gulp.emit('task_start', metadata);
  try {
    // @ts-ignore
    taskInstance.apply(gulp);

    metadata.hrDuration = process.hrtime(start);
    gulp.emit('task_stop', metadata);
    gulp.emit('stop');
  } catch (err) {
    err.hrDuration = process.hrtime(start);
    err.task = metadata.task;
    gulp.emit('task_err', err);
  }
}
