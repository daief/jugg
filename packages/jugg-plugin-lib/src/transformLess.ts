const less = require('less');
const { readFileSync } = require('fs');
const path = require('path');
const postcss = require('postcss');
const NpmImportPlugin = require('less-plugin-npm-import');
const rucksack = require('rucksack-css');
const autoprefixer = require('autoprefixer');

export default function transformLess(lessFile: string, config: any = {}): Promise<any> {
  const { cwd = process.cwd() } = config;
  const resolvedLessFile = path.resolve(cwd, lessFile);

  let data = readFileSync(resolvedLessFile, 'utf-8');
  data = data.replace(/^\uFEFF/, '');

  // Do less compile
  const lessOpts = {
    paths: [path.dirname(resolvedLessFile)],
    filename: resolvedLessFile,
    plugins: [new NpmImportPlugin({ prefix: '~' })],
    javascriptEnabled: true,
  };
  return (
    less
      .render(data, lessOpts)
      // @ts-ignore
      .then(result =>
        postcss([
          rucksack(),
          autoprefixer({
            browsers: [
              'last 2 versions',
              'Firefox ESR',
              '> 1%',
              'ie >= 9',
              'iOS >= 8',
              'Android >= 4',
            ],
          }),
        ]).process(result.css, { from: undefined })
      )
      // @ts-ignore
      .then(r => r.css)
  );
}
