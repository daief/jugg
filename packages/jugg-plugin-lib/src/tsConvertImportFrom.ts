/**
 * adjust `import` and `require` statement, convert `match` to `target`
 * can use `// jugg-lib-disable` to disable
 * @example
 *    opt = { match: /\.vue$/, target: '.js' }
 *
 *    import A from '../A.vue'
 *    import A from '../A.js'
 *
 *    require('A.vue')
 *    require('A.js')
 *
 *    require('A.vue') // jugg-lib-disable
 *    require('A.vue')
 */

import * as ts from 'typescript';

export interface IOptions {
  match?: string | RegExp;
  target?: string;
}

export function createTransformer(opts: IOptions = {}) {
  const { match = /\.vue$/i, target = '.js' } = opts;

  const createNewStringLiteralByOpts = (node: ts.StringLiteral) => {
    return ts.createStringLiteral(node.text.replace(match, target));
  };

  const transformer: ts.TransformerFactory<ts.SourceFile> = context => {
    return transformerNode => {
      const hasDisableComment = (node: ts.Node) => {
        const range = ts.getTrailingCommentRanges(transformerNode.text, node.end) || [];

        return (
          range
            .map(ran => {
              return transformerNode.text.substring(ran.pos, ran.end);
            })
            // use `jugg-lib-disable` to disable converting
            .filter(comment => /\/\/\s*jugg-lib-disable/i.test(comment)).length > 0
        );
      };

      const visitorImport: ts.Visitor = node => {
        if (ts.isImportDeclaration(node) || ts.isExportDeclaration(node)) {
          const importFrom = node.moduleSpecifier as ts.StringLiteral;
          if (!importFrom || hasDisableComment(node)) {
            return node;
          }
          const newImportFrom = createNewStringLiteralByOpts(importFrom);
          return ts.isImportDeclaration(node)
            ? ts.updateImportDeclaration(
                node,
                node.decorators,
                node.modifiers,
                node.importClause,
                newImportFrom
              )
            : ts.updateExportDeclaration(
                node,
                node.decorators,
                node.modifiers,
                node.exportClause,
                newImportFrom
              );
        }
        return ts.visitEachChild(node, visitorImport, context);
      };

      const visitRequire: ts.Visitor = node => {
        if (ts.isExpressionStatement(node) && ts.isCallExpression(node.expression)) {
          /**
           * make is `require('..')`
           */
          const callExpression = node.expression;
          const expression = callExpression.expression;
          const args = callExpression.arguments;
          const requireStr = args[0];
          if (
            !ts.isIdentifier(expression) ||
            expression.text !== 'require' ||
            args.length !== 1 ||
            !ts.isStringLiteral(requireStr) ||
            hasDisableComment(node)
          ) {
            return node;
          }

          return ts.updateExpressionStatement(
            node,
            ts.createIdentifier(`require("${createNewStringLiteralByOpts(requireStr).text}")`)
          );
        }
        return ts.visitEachChild(node, visitRequire, context);
      };

      return ts.visitNode(ts.visitNode(transformerNode, visitorImport), visitRequire);
    };
  };
  return transformer;
}

export default createTransformer;
