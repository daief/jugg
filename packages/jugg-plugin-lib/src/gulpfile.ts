import gulp from 'gulp';
import { PluginAPI } from '@axew/jugg/types/PluginAPI';
import through2 from 'through2';
import transformLess from './transformLess';
import gulpTs from 'gulp-typescript';
import merge2 from 'merge2';
import gulpVue from './gulpVue';
import { ScriptTarget, ModuleResolutionKind, ModuleKind } from 'typescript';
import transformerFactory from './tsConvertImportFrom';

export interface IOptions {
  /**
   * convert less import in es/lib to css file path, default `true`
   */
  convertLessImport2Css?: boolean;
  /**
   * copy file to dest with this suffix, built-in `png|jpg|jpeg|gif|webp|svg|mp4|webm|ogg|mp3|wav|flac|aac|woff|woff2|eot|ttf|otf`
   * @example `md|html`
   */
  copyFileSuffix?: string;
}

export default (opts: IOptions, api: PluginAPI) => {
  const { convertLessImport2Css = true, copyFileSuffix = '' } = opts;

  const appendSuffix = copyFileSuffix.replace(/^\|*/, '').replace(/\|*$/, '');

  const { getAbsolutePath, logger } = api.jugg.Utils;
  const LIB_DIR = getAbsolutePath('lib');
  const ES_DIR = getAbsolutePath('es');

  function compile(modules: boolean) {
    const TARGET_DIR = modules === false ? ES_DIR : LIB_DIR;
    const less = gulp
      .src(['src/**/*.less'])
      .pipe(
        through2.obj(function(file, _2, next) {
          this.push(file.clone());

          transformLess(file.path)
            .then(css => {
              file.contents = Buffer.from(css);
              file.path = file.path.replace(/\.less$/, '.css');
              this.push(file);
              next();
            })
            .catch(e => {
              logger.error(e);
            });
        })
      )
      .pipe(gulp.dest(TARGET_DIR));

    // copy files
    const assets = gulp
      .src([
        `src/**/*.@(png|jpg|jpeg|gif|webp|svg|mp4|webm|ogg|mp3|wav|flac|aac|woff|woff2|eot|ttf|otf${
          appendSuffix ? '|' + appendSuffix : ''
        })`,
      ])
      .pipe(gulp.dest(TARGET_DIR));

    const BASE_COMPILER_OPTIONS_FN = (isGulpTs = true): any => {
      const COMMON_CFG = {
        noUnusedParameters: true,
        noUnusedLocals: true,
        strictNullChecks: true,
        allowSyntheticDefaultImports: true,
      };
      return isGulpTs === true
        ? {
            ...COMMON_CFG,
            target: 'es5',
            moduleResolution: 'node',
            module: modules ? 'commonjs' : 'esnext',
          }
        : {
            ...COMMON_CFG,
            target: ScriptTarget.ES5,
            moduleResolution: ModuleResolutionKind.NodeJs,
            module: modules ? ModuleKind.CommonJS : ModuleKind.ESNext,
          };
    };

    const compileTS = (src: string[], tsOpts: any = {}) => {
      const tsProject = gulpTs.createProject(getAbsolutePath('tsconfig.json'), {
        ...BASE_COMPILER_OPTIONS_FN(),
        ...tsOpts,
        getCustomTransformers: () => ({
          before: [transformerFactory()],
        }),
      });

      const rs = gulp.src(src).pipe(tsProject(gulpTs.reporter.fullReporter()));

      return rs;
    };

    // use tsc compile ts, tsx, with declaration files
    const tsResult = compileTS(['src/**/*.@(ts|tsx)'], {
      allowJs: false,
      declaration: true,
    });

    // use tsc compile js, jsx, no declaration files
    const tsJsResult = compileTS(['src/**/*.@(js|jsx)'], {
      allowJs: true,
      declaration: false,
    });

    const vueResult = gulp
      .src('src/**/*.vue')
      .pipe(
        gulpVue({
          tsCompilerOptions: {
            ...BASE_COMPILER_OPTIONS_FN(false),
          },
        })
      )
      .pipe(gulp.dest(TARGET_DIR));

    const convertLessImport2CssStream = () =>
      through2.obj(function z(file, encoding, next) {
        this.push(file.clone());
        if (file.path.match(/(\/|\\)style(\/|\\)index\.js/) && convertLessImport2Css === true) {
          const content = file.contents.toString(encoding);
          const cssInjection = (c: string) =>
            c
              .replace(/\/style\/?'/g, `/style/css'`)
              .replace(/\/style\/?"/g, `/style/css"`)
              .replace(/\.less/g, '.css');

          file.contents = Buffer.from(cssInjection(content));
          file.path = file.path.replace(/index\.js/, 'css.js');
          this.push(file);
          next();
        } else {
          next();
        }
      });

    return merge2([
      less,
      tsResult.js.pipe(convertLessImport2CssStream()).pipe(gulp.dest(TARGET_DIR)),
      tsResult.dts.pipe(gulp.dest(TARGET_DIR)),
      tsJsResult.js.pipe(convertLessImport2CssStream()).pipe(gulp.dest(TARGET_DIR)),
      assets,
      vueResult,
    ]);
  }

  gulp.task('compile-with-es', done => {
    logger.info('[Parallel] Compile to es...');
    compile(false).on('finish', done);
  });

  gulp.task('compile-with-lib', done => {
    logger.info('[Parallel] Compile to js...');
    compile(true).on('finish', done);
  });

  gulp.task('compile', gulp.parallel('compile-with-es', 'compile-with-lib'));

  return gulp;
};
