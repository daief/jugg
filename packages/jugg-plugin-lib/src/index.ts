import { PluginAPI } from '@axew/jugg/types/PluginAPI';
import { runTask } from './runTask';
import gulpfile from './gulpfile';

export default function build(api: PluginAPI) {
  api.registerCommand({
    command: 'lib',
    description: 'build lib',
    action: () => {
      gulpfile({}, api);
      runTask('compile');
    },
  });
}
