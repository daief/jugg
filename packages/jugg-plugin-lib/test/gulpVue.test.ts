import * as fs from 'fs';
import * as path from 'path';
import { compileVueFile } from '../src/gulpVue';

const VUE_DIR = path.resolve(__dirname, 'vue-files');

describe('compile-vue-file', () => {
  fs.readdirSync(VUE_DIR).forEach(filename => {
    it(`compile [${filename}]`, () => {
      const fullpath = path.resolve(VUE_DIR, filename);
      const result = compileVueFile(
        fs.readFileSync(fullpath, 'utf-8'),
        filename,
      );
      expect(result).toMatchSnapshot();
    });
  });
});
