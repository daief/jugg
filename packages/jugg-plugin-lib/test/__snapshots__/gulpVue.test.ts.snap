// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile-vue-file compile [a.vue] 1`] = `
"/* script */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === \\"object\\" && typeof Reflect.decorate === \\"function\\") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Component, Vue } from 'vue-property-decorator';
let A = class A extends Vue {
};
A = __decorate([
    Component
], A);
const __vue_script__ = A;

/* template */
var __vue_render__ = function() {
  var _vm = this
  var _h = _vm.$createElement
  var _c = _vm._self._c || _h
  return _c(\\"div\\")
}
var __vue_staticRenderFns__ = []
__vue_render__._withStripped = true

  /* style */
  const __vue_inject_styles__ = undefined
  /* scoped */
  const __vue_scope_id__ = undefined
  /* module identifier */
  const __vue_module_identifier__ = undefined
  /* functional template */
  const __vue_is_functional_template__ = false
  /* component normalizer */
  import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.js'
  /* style inject */
  
  /* style inject SSR */
  

  
  export default __vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    undefined,
    undefined
  )"
`;

exports[`compile-vue-file compile [b.vue] 1`] = `
"/* script */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === \\"object\\" && typeof Reflect.decorate === \\"function\\") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Component, Vue } from 'vue-property-decorator';
// @ts-ignore
import Custom from 'xx-lib/Custom.vue';
let default_1 = class extends Vue {
};
default_1 = __decorate([
    Component({
        components: {
            Custom,
        },
    })
], default_1);
const __vue_script__ = default_1;

/* template */
var __vue_render__ = function() {
  var _vm = this
  var _h = _vm.$createElement
  var _c = _vm._self._c || _h
  return _c(\\"div\\", [_c(\\"Custom\\")], 1)
}
var __vue_staticRenderFns__ = []
__vue_render__._withStripped = true

  /* style */
  const __vue_inject_styles__ = undefined
  /* scoped */
  const __vue_scope_id__ = \\"data-v-7dc80285\\"
  /* module identifier */
  const __vue_module_identifier__ = undefined
  /* functional template */
  const __vue_is_functional_template__ = false
  /* component normalizer */
  import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.js'
  /* style inject */
  
  /* style inject SSR */
  

  
  export default __vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    undefined,
    undefined
  )"
`;
